### PR TITLE
Use exceptions instead of return values to handle errors. Implemented unit tests.

### PR DESCRIPTION
**Pull request by cnv1989**
In this PR, I am trying to fix some function definitions to return values instead of using pointers to pass the result. Some organizational changes to improve readability and added unit tests.

I updated CMakelists.txt to download a model from S3 bucket. Please double check this to ensure there are not unknown side effects.

Thanks for contributing to DLR! By submitting this pull request, you confirm that your contribution is made under the terms of the [Apache 2.0 license](https://github.com/neo-ai/neo-ai-dlr/blob/master/LICENSE).

Please refer to our [guideline](https://github.com/neo-ai/neo-ai-dlr/blob/master/CONTRIBUTING.md) for useful information and tips.

### CODE REVIEW

File: include/dlr_common.h [hunk 1]
```diff
@@ -86,51 +86,47 @@ class DLRModel {
   DLContext ctx_;
   std::vector<std::string> input_names_;
   std::vector<std::string> input_types_;
+  std::vector<int64_t> output_shape_;
```
Comment (cnv1989) [inline include/dlr_common.h:89]: Not used delete it.

File: src/dlr_tvm.cc [hunk 2]
```diff
@@ -257,13 +253,10 @@ int TVMModel::GetOutputIndex(const char* name) const {
       return i;
     }
   }
-  return -1;
+  throw dmlc::Error("Couldn't find index for output node!");
```
Comment (trevor-m) [inline src/dlr_tvm.cc:256]: Can we also print the name?

File: tests/cpp/dlr_tvm_test.cc [hunk 3]
```diff
@@ -1,8 +1,136 @@
...
+  } catch(const dmlc::Error& e) {
+    EXPECT_STREQ(e.what(), "Couldn't find index for output node!");
+  }
+  
+  try {
+    model.GetOutputByName("blah", output1);
+  } catch(const dmlc::Error& e) {
+    EXPECT_STREQ(e.what(), "Couldn't find index for output node!");
+  }
+}
```
Comment (trevor-m) [inline tests/cpp/dlr_tvm_test.cc:55]: Can you also test GetOutput() where index is greater than the number of outputs?

File: tests/cpp/test_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  LOG(INFO) << "Image read - OK, float[" << i << "]";
+
+  for (int j = 1; j < batch_size; j++) {
+    std::memcpy(img + j * size, img, size * sizeof(float));
+  }
+  return img;
+}
+
+
+#endif
```
Comment (trevor-m) [inline tests/cpp/test_utils.hpp:34]: Add newline at end of file

File: tests/cpp/dlr_tvm_test.cc (hunk 3 above)
Comment (cnv1989) [reply tests/cpp/dlr_tvm_test.cc:61]: 👍

File: src/dlr_tvm.cc [hunk 5]
```diff
@@ -228,8 +227,7 @@ bool TVMModel::HasMetadata() const { return !this->metadata.is_null(); }
 
 const char* TVMModel::GetOutputName(const int index) const {
   if (!this->HasMetadata()) {
-    LOG(INFO) << "No metadata file was found!";
-    return nullptr;
+    throw dmlc::Error("No metadata file was found!");
```
Comment (trevor-m) [inline src/dlr_tvm.cc:230]: What happens during `_fetch_output_names()` when there is no metadata file?

Comment (cnv1989) [reply src/dlr_tvm.cc:230]: We return -1 in the C APIs which will throw an error in python bindings.

Comment (cnv1989) [reply src/dlr_tvm.cc:230]: https://github.com/neo-ai/neo-ai-dlr/blob/ef33b6687bae6e084fa077b51bc45efaa054318b/python/dlr/dlr_model.py#L212

**Pull request merged**